lib/elftools/constants.rb

```diff
@@ -42,120 +42,371 @@ module DF
-      DT_BIND_NOW     = 24 # if the loader needs to do relocate now, superseded by the DF_BIND_NOW flag
-      DT_INIT_ARRAY   = 25 # address init array
-      DT_FINI_ARRAY   = 26 # address of fini array
-      DT_INIT_ARRAYSZ = 27 # total size of init array
-      DT_FINI_ARRAYSZ = 28 # total size of fini array
-      DT_RUNPATH      = 29 # path of libraries for searching
-      DT_FLAGS        = 30 # flags
-      DT_ENCODING     = 32 # just a lower bound
+      DT_NULL                       = 0 # marks the end of the _DYNAMIC array
+      DT_NEEDED                     = 1 # libraries need to be linked by loader
+      DT_PLTRELSZ                   = 2 # total size of relocation entries
+      DT_PLTGOT                     = 3 # address of procedure linkage table or global offset table
+      DT_HASH                       = 4 # address of symbol hash table
+      DT_STRTAB                     = 5 # address of string table
+      DT_SYMTAB                     = 6 # address of symbol table
+      DT_RELA                       = 7 # address of a relocation table
+      DT_RELASZ                     = 8 # total size of the {DT_RELA} table
+      DT_RELAENT                    = 9 # size of each entry in the {DT_RELA} table
+      DT_STRSZ                      = 10 # total size of {DT_STRTAB}
+      DT_SYMENT                     = 11 # size of each entry in {DT_SYMTAB}
+      DT_INIT                       = 12 # where the initialization function is
+      DT_FINI                       = 13 # where the termination function is
+      DT_SONAME                     = 14 # the shared object name
+      DT_RPATH                      = 15 # has been superseded by {DT_RUNPATH}
+      DT_SYMBOLIC                   = 16 # has been superseded by the DF_SYMBOLIC flag
+      DT_REL                        = 17 # similar to {DT_RELA}
+      DT_RELSZ                      = 18 # total size of the {DT_REL} table
+      DT_RELENT                     = 19 # size of each entry in the {DT_REL} table
+      DT_PLTREL                     = 20 # type of relocation entry, either {DT_REL} or {DT_RELA}
+      DT_DEBUG                      = 21 # for debugging
+      DT_TEXTREL                    = 22 # has been superseded by the DF_TEXTREL flag
+      DT_JMPREL                     = 23 # address of relocation entries associated solely with procedure linkage table
+      DT_BIND_NOW                   = 24 # if the loader needs to do relocate now, superseded by the DF_BIND_NOW flag
+      DT_INIT_ARRAY                 = 25 # address init array
+      DT_FINI_ARRAY                 = 26 # address of fini array
+      DT_INIT_ARRAYSZ               = 27 # total size of init array
+      DT_FINI_ARRAYSZ               = 28 # total size of fini array
+      DT_RUNPATH                    = 29 # path of libraries for searching
+      DT_FLAGS                      = 30 # flags
+      DT_ENCODING                   = 32 # just a lower bound
```

Comment:
Seems can be removed as it has the same value as DT_PREINIT_ARRAY. Is it still used in binutils?

Reply:
`DT_ENCODING` isn't a real enum entry but a special value:

> A tag whose value is an even number indicates a dynamic section entry that uses `d_ptr`. A tag whose value is an odd number indicates a dynamic section entry that uses `d_val` or that uses neither `d_ptr` nor `d_val`. Tags whose values are less than the special value `DT_ENCODING` and tags whose values fall between `DT_HIOS` and `DT_LOPROC` do not follow these rules. 

Looking at the the values after:

```
DT_PREINIT_ARRAY -> even -> d_ptr -> correct given it's an address
DT_PREINIT_ARRAYSZ -> odd -> d_val -> correct given it's a size
DT_SYMTAB_SHNDX -> even -> d_ptr -> correct given it's an address
DT_RELRSZ -> odd -> d_val -> correct given it's a size
DT_RELR -> even -> d_ptr -> correct given it's an address
DT_RELRENT -> odd -> d_val -> correct given it's a count
```

When `DT_RELRENT` was added, `DT_ENCODING` did get bumped in bintuils. This change is however disputed (it hasn't changed anywhere else like LLVM, glibc or even elfcpp from binutils) and I do not believe the change was correct.

Reply:
Thanks for the detailed explanation!